src/markov/mc_tools.jl

```diff
@@ -57,172 +59,217 @@ end
+- `mc::MarkovChain` : MarkovChain instance.
 
 ##### Returns
 
-- `x::Vector{Vector}`: A `Vector` containing `Vector{Int}`s that describe the
-recurrent classes of the transition matrix for p
+- `::Vector{Vector{Int}}` : Vector of vectors that describe the recurrent
+classes of `mc`.
 
 """
 recurrent_classes(mc::MarkovChain) = attracting_components(DiGraph(mc.p))
 
 """
-A communication class of the Markov Chain `X_t` or of the stochastic matrix `p`
-is a strongly connected component of the directed graph associated with `p`
+Find the communication classes of the Markov chain `mc`.
 
 #### Arguments
 
-- `mc::MarkovChain` MarkovChain instance containing a valid stochastic matrix
+- `mc::MarkovChain` : MarkovChain instance.
 
 ### Returns
 
-- `x::Vector{Vector{Int64}}` An array of the associated strongly connected components
+- `::Vector{Vector{Int}}` : Vector of vectors that describe the communication
+classes of `mc`.
 
 """
 communication_classes(mc::MarkovChain) = strongly_connected_components(DiGraph(mc.p))
 
 """
-Indicate whether the Markov chain is irreducible.
+Indicate whether the Markov chain `mc` is irreducible.
 
 ##### Arguments
 
-- `mc::MarkovChain` : MarkovChain instance containing a valid stochastic matrix
+- `mc::MarkovChain` : MarkovChain instance.
 
 ##### Returns
 
-- Bool (true or false)
+- `::Bool`
 
 """
 is_irreducible(mc::MarkovChain) =  is_strongly_connected(DiGraph(mc.p))
 
 """
-Indicate whether the Markov chain is aperiodic.
+Indicate whether the Markov chain `mc` is aperiodic.
 
 ##### Arguments
 
-- `mc::MarkovChain` : MarkovChain instance containing a valid stochastic matrix
+- `mc::MarkovChain` : MarkovChain instance.
 
 ##### Returns
 
-- Bool (true or false)
+- `::Bool`
 
 """
 is_aperiodic(mc::MarkovChain) = period(mc) == 1
 
+"""
+Return the period of the Markov chain `mc`.
+
+##### Arguments
+
+- `mc::MarkovChain` : MarkovChain instance.
+
+##### Returns
+
+- `::Int` : Period of `mc`.
+
+"""
 function period(mc::MarkovChain)
     g = DiGraph(mc.p)
     recurrent = attracting_components(g)
-    periods   = Int[]
 
+    d = 1
     for r in recurrent
-        push!(periods, period(g[r]))
+        pd = period(g[r])
+        d *= div(pd, gcd(pd, d))
     end
-    return lcm(periods)
+
+    return d
 end
-"""
-calculate the stationary distributions associated with a N-state markov chain
+
+
+for (S, ex_T, ex_gth) in ((Real, :(T), :(gth_solve!)),
+                          (Integer, :(Rational{T}), :(gth_solve)))
+    @eval function stationary_distributions{T<:$S}(mc::MarkovChain{T})
+        n = n_states(mc)
+        rec_classes = recurrent_classes(mc)
+        T1 = $ex_T
+        stationary_dists = Array(Vector{T1}, length(rec_classes))
+
+        for (i, rec_class) in enumerate(rec_classes)
+            dist = zeros(T1, n)
+            # * mc.p[rec_class, rec_class] is a copy, hence gth_solve!
+            #   while gth_solve for Int matrix
+            # * full is to convert a sparse matrix to a dense matrix
```

Comment:
A sparse matrix `A` has to be converted a dense matrix `B`. If the eltype of `A` is `Integer`, it has to be converted to `Rational` so that the eltype of `B` is `Rational`. We could write a custom function that does this.